cpp/src/neighbors/detail/cagra/cagra_build.cuh

```diff
@@ -445,38 +442,183 @@ index<T, IdxT> build(
     }
   }
 
+  auto cagra_graph = raft::make_host_matrix<IdxT, int64_t>(0, 0);
+
   // Dispatch based on graph_build_params
-  if (std::holds_alternative<cagra::graph_build_params::ivf_pq_params>(knn_build_params)) {
-    auto ivf_pq_params =
-      std::get<cuvs::neighbors::cagra::graph_build_params::ivf_pq_params>(knn_build_params);
-    build_knn_graph(res, dataset, knn_graph->view(), ivf_pq_params);
-  } else {
-    auto nn_descent_params =
-      std::get<cagra::graph_build_params::nn_descent_params>(knn_build_params);
+  if (std::holds_alternative<cagra::graph_build_params::iterative_search_params>(
+        knn_build_params)) {
+    // Iteratively improve the accuracy of the graph by repeatedly running
+    // CAGRA's search() and optimize(). As for the size of the graph, instead
+    // of targeting all nodes from the beginning, the number of nodes is
+    // initially small, and the number of nodes is doubled with each iteration.
+    RAFT_LOG_INFO(
+      "Iteratively creating/improving graph index using CAGRA's search() and optimize()");
+
+    // If dataset is a host matrix, change it to a device matrix. Also, if the
+    // dimensionality of the dataset does not meet the alighnemt restriction,
+    // add extra dimensions and change it to a strided matrix.
+    auto dev_aligned_dataset      = make_aligned_dataset(res, dataset);
```

Comment:
This will OOM for large dataset, could we throw an informative error message in that case?
```suggestion
  std::unique_ptr<strided_dataset<T, int64_t>> dev_aligned_dataset;
try {
  dev_aligned_dataset = make_aligned_dataset(res, dataset);
} catch (raft::logic_error& e) {
  RAFT_LOG_ERROR("Iterative CAGRA build requires the dataset to fit GPU memory");
  throw e;
}
```

Reply:
Thanks, will fix in that way as well.